File: merge_sort/merge_sort.c
#include <time.h>
#include <err.h>
#include <linux/perf_event.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdint.h>
#include <inttypes.h>

#define TOTAL_EVENTS 5

static long perf_event_open(struct perf_event_attr *hw_event, pid_t pid,
                            int cpu, int group_fd, unsigned long flags)
{
    return syscall(SYS_perf_event_open, hw_event, pid, cpu, group_fd, flags);
}

void configure_event(struct perf_event_attr *pe, uint32_t type, uint64_t config)
{
    memset(pe, 0, sizeof(struct perf_event_attr));
    pe->type = type;
    pe->size = sizeof(struct perf_event_attr);
    pe->config = config;
    pe->read_format = PERF_FORMAT_GROUP | PERF_FORMAT_ID;
    pe->disabled = 1;
    pe->exclude_kernel = 1;
    pe->exclude_hv = 1;
}

struct read_format
{
    uint64_t nr;
    struct
    {
        uint64_t value;
        uint64_t id;
    } values[TOTAL_EVENTS];
};

void merge(int arr[], int left, int mid, int right)
{
    int i, j, k;
    int n1 = mid - left + 1;
    int n2 = right - mid;

    int *leftArr = (int *)malloc(n1 * sizeof(int));
    int *rightArr = (int *)malloc(n2 * sizeof(int));

    for (i = 0; i < n1; i++)
        leftArr[i] = arr[left + i];
    for (j = 0; j < n2; j++)
        rightArr[j] = arr[mid + 1 + j];

    i = 0;
    j = 0;
    k = left;
    while (i < n1 && j < n2)
    {
        if (leftArr[i] <= rightArr[j])
        {
            arr[k] = leftArr[i];
            i++;
        }
        else
        {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    while (i < n1)
    {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    while (j < n2)
    {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
    free(leftArr);
    free(rightArr);
}

void merge_sort(int arr[], int left, int right)
{
    if (left < right)
    {
        int mid = left + (right - left) / 2;
        merge_sort(arr, left, mid);
        merge_sort(arr, mid + 1, right);

        merge(arr, left, mid, right);
    }
}

double diff_timespec(struct timespec *time1, struct timespec *time0)
{
    return (time1->tv_sec - time0->tv_sec) + (time1->tv_nsec - time0->tv_nsec) / 1000000000.0;
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: %s <array_size>\n", argv[0]);
        return 1;
    }

    int size = atoi(argv[1]);
    if (size <= 0)
    {
        printf("Error: array size must be positive integers.\n");
        return 1;
    }

    struct timespec start_time, end_time;
    int *arr = (int *)malloc(size * sizeof(int));
    if (!arr)
    {
        printf("Error: Memory allocation failed.\n");
        return 1;
    }

    srand(10000);
    for (int i = 0; i < size; i++)
    {
        arr[i] = rand();
    }

    int fd[TOTAL_EVENTS];
    int id[TOTAL_EVENTS];
    uint64_t pe_val[TOTAL_EVENTS];
    struct perf_event_attr pe[TOTAL_EVENTS];
    struct read_format counter_results;

    // Configure desired events
    configure_event(&pe[0], PERF_TYPE_HARDWARE, PERF_COUNT_HW_CPU_CYCLES);
    configure_event(&pe[1], PERF_TYPE_HARDWARE, PERF_COUNT_HW_INSTRUCTIONS);
    configure_event(&pe[2], PERF_TYPE_HW_CACHE,
                    (PERF_COUNT_HW_CACHE_L1D | (PERF_COUNT_HW_CACHE_OP_READ << 8) |
                     (PERF_COUNT_HW_CACHE_RESULT_MISS << 16)));
    configure_event(&pe[3], PERF_TYPE_HW_CACHE,
                    (PERF_COUNT_HW_CACHE_LL | (PERF_COUNT_HW_CACHE_OP_READ << 8) |
                     (PERF_COUNT_HW_CACHE_RESULT_MISS << 16)));
    configure_event(&pe[4], PERF_TYPE_HARDWARE, PERF_COUNT_HW_BRANCH_MISSES);

    // Open group leader
    fd[0] = perf_event_open(&pe[0], 0, -1, -1, 0);
    ioctl(fd[0], PERF_EVENT_IOC_ID, &id[0]);

    // Open other events as part of the group
    for (int i = 1; i < TOTAL_EVENTS; i++)
    {
        fd[i] = perf_event_open(&pe[i], 0, -1, fd[0], 0);
        ioctl(fd[i], PERF_EVENT_IOC_ID, &id[i]);
    }

    ioctl(fd[0], PERF_EVENT_IOC_RESET, PERF_IOC_FLAG_GROUP);
    ioctl(fd[0], PERF_EVENT_IOC_ENABLE, PERF_IOC_FLAG_GROUP);

    clock_gettime(CLOCK_MONOTONIC, &start_time);
    merge_sort(arr, 0, size - 1);
    clock_gettime(CLOCK_MONOTONIC, &end_time);

    ioctl(fd[0], PERF_EVENT_IOC_DISABLE, PERF_IOC_FLAG_GROUP);
    read(fd[0], &counter_results, sizeof(struct read_format));

    printf("Num events captured: %" PRIu64 "\n", counter_results.nr);
    for (int i = 0; i < counter_results.nr; i++)
    {
        for (int j = 0; j < TOTAL_EVENTS; j++)
        {
            if (counter_results.values[i].id == id[j])
            {
                pe_val[j] = counter_results.values[i].value;
            }
        }
    }

    printf("CPU cycles:             %" PRIu64 "\n", pe_val[0]);
    printf("Instructions retired:   %" PRIu64 "\n", pe_val[1]);
    printf("L1 DCache read misses:   %" PRIu64 "\n", pe_val[2]);
    printf("Last level DCache read misses:   %" PRIu64 "\n", pe_val[3]);
    printf("Branch misses:           %" PRIu64 "\n", pe_val[4]);
    printf("IPC (instructions/cycle): %.2f\n", (double)pe_val[1] / pe_val[0]);

    for (int i = 0; i < TOTAL_EVENTS; i++)
    {
        close(fd[i]);
    }

    double time_elapsed = diff_timespec(&end_time, &start_time);
    printf("Time taken to sort the array of size %d: %f seconds\n", size, time_elapsed);

    free(arr);
    return 0;
}

File: merge_sort/src/merge_sort.rs
use std::env;
use std::process::exit;
use libc::{clock_gettime, timespec, CLOCK_MONOTONIC, srand, rand};

fn merge<T: Ord + Copy>(arr: &mut [T], mid: usize) {
    let mut left = arr[..mid].to_vec();
    let mut right = arr[mid..].to_vec();

    let (mut l, mut r, mut i) = (0, 0, 0);

    while l < left.len() && r < right.len() {
        if left[l] <= right[r] {
            arr[i] = left[l];
            l += 1;
        } else {
            arr[i] = right[r];
            r += 1;
        }
        i += 1;
    }

    while l < left.len() {
        arr[i] = left[l];
        l += 1;
        i += 1;
    }

    while r < right.len() {
        arr[i] = right[r];
        r += 1;
        i += 1;
    }
}

pub fn merge_sort<T: Ord + Copy>(arr: &mut [T]) {
    if arr.len() > 1 {
        let mid = arr.len() / 2;
        merge_sort(&mut arr[..mid]);
        merge_sort(&mut arr[mid..]);
        merge(arr, mid);
    }
}


fn diff_timespec(time1: &timespec, time0: &timespec) -> f64 {
    (time1.tv_sec - time0.tv_sec) as f64 +
        (time1.tv_nsec - time0.tv_nsec) as f64 / 1_000_000_000.0
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        eprintln!("Usage: {} <array_size>", args[0]);
        exit(1);
    }

    let size: usize = match args[1].parse() {
        Ok(n) if n > 0 => n,
        _ => {
            eprintln!("Error: array size must be a positive integer.");
            exit(1);
        }
    };

    unsafe {
        srand(10000);
    }

    let mut arr: Vec<i32> = (0..size).map(|_| unsafe { rand() }).collect();

    let mut start_time: timespec = timespec { tv_sec: 0, tv_nsec: 0 };
    unsafe {
        clock_gettime(CLOCK_MONOTONIC, &mut start_time);
    }

    merge_sort(&mut arr);

    let mut end_time: timespec = timespec { tv_sec: 0, tv_nsec: 0 };
    unsafe {
        clock_gettime(CLOCK_MONOTONIC, &mut end_time);
    }

    let time_elapsed = diff_timespec(&end_time, &start_time);

    unsafe {
        println!(
            "Time taken to sort the array of size {}: {:.6} seconds",
            size,
            time_elapsed
        );
    }
}